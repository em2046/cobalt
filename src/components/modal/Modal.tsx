import ClassNameBuilder from '../../utils/class-name-builder'
import * as React from 'react'
import * as ReactDOM from 'react-dom'
import Button from '../button'
import Icon from '../icon'
import { useEffect } from 'react'

interface ModalBaseProps {
  ariaId?: string
  title?: string
  mask?: boolean
  maskClosable?: boolean
  onCancel?: React.MouseEventHandler
  onOk?: React.MouseEventHandler
}

export interface ModalProps extends ModalBaseProps {
  children?: React.ReactNode
  visible?: boolean
  getContainer?: () => HTMLElement | undefined
  footer?: React.ReactNode
  closable?: boolean
  wrapClassName?: string
}

export interface ModalFuncProps extends ModalBaseProps {
  content?: React.ReactNode
}

export const prefixClass = 'cobalt-modal'

let dialogCounter = 1

export default function Modal(props: ModalProps) {
  useEffect(() => {
    dialogCounter++
  }, [])

  let {
    ariaId,
    title,
    children,
    mask = true,
    visible,
    onCancel,
    onOk,
    maskClosable = true,
    getContainer = () => document.body,
    footer,
    wrapClassName,
    closable = true
  } = props

  let classNameBuilder = new ClassNameBuilder(prefixClass, wrapClassName)
  let classes = classNameBuilder.toString()

  let wrapClass = `${prefixClass}-wrap`
  let maskClass = `${prefixClass}-mask`
  let headerClass = `${prefixClass}-header`
  let bodyClass = `${prefixClass}-body`
  let footerClass = `${prefixClass}-footer`
  let titleClass = `${prefixClass}-title`
  let closeClass = `${prefixClass}-close`

  let hasHeader = closable || title
  let footerIsNull = footer === null

  ariaId = ariaId || dialogCounter.toString()
  let labelledby = hasHeader ? `dialog${ariaId}Title` : undefined
  let describedby = `dialog${ariaId}Desc`

  let maskNode = mask ? <div className={maskClass} /> : null

  if (!visible) {
    return null
  }

  function handleMaskClick(e: React.MouseEvent) {
    if (onCancel && maskClosable) {
      onCancel(e)
    }
  }

  function handleModalClick(e: React.MouseEvent) {
    e.stopPropagation()
  }

  function handleCancel(e: React.MouseEvent) {
    if (onCancel) {
      onCancel(e)
    }
  }

  function handleOk(e: React.MouseEvent) {
    if (onOk) {
      onOk(e)
    }
  }

  let modalClose = closable ? (
    <button className={closeClass} onClick={handleCancel}>
      <Icon type="close" />
    </button>
  ) : null

  let modalHeader = hasHeader ? (
    <div className={headerClass}>
      <div id={labelledby} className={titleClass}>
        {title}
      </div>
      {modalClose}
    </div>
  ) : null

  let modalBody = (
    <div id={describedby} className={bodyClass}>
      {children}
    </div>
  )

  let defaultFooter = (
    <div className={footerClass}>
      <Button onClick={handleCancel}>Cancel</Button>
      <Button type="primary" onClick={handleOk}>
        OK
      </Button>
    </div>
  )

  let customFooter = <div className={footerClass}>{footer}</div>
  let computedFooter = footer ? customFooter : defaultFooter
  let modalFooter = footerIsNull ? null : computedFooter

  let dialog = (
    <div>
      {maskNode}
      <div
        role="dialog"
        aria-labelledby={labelledby}
        aria-describedby={describedby}
        className={wrapClass}
        onClick={handleMaskClick}
      >
        <div className={classes} onClick={handleModalClick}>
          {modalHeader}
          {modalBody}
          {modalFooter}
        </div>
      </div>
    </div>
  )

  let container = getContainer() || document.body
  return ReactDOM.createPortal(dialog, container)
}
