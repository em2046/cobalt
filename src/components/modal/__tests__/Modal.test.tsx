import React from 'react';
import renderer from 'react-test-renderer';
import Modal, { ModalProps } from '../Modal';

jest.mock('react-dom', () => ({
  createPortal: (node: React.ReactNode) => node
}));

interface State {
  visible: boolean;
}

class ModalTester extends React.Component<ModalProps, State> {
  constructor(props: ModalProps) {
    super(props);
    this.state = {
      visible: false
    };
  }

  container?: HTMLElement;

  componentWillMount(): void {
    this.setState({
      visible: true
    });
  }

  saveContainer = (container: HTMLDivElement) => {
    this.container = container;
  };

  getContainer = () => {
    return this.container;
  };

  render(): React.ReactNode {
    const { visible } = this.state;
    return (
      <div>
        <div ref={this.saveContainer} />
        <Modal
          {...this.props}
          visible={visible}
          getContainer={this.getContainer}
        >
          Here is content of Modal
        </Modal>
      </div>
    );
  }
}

describe('Modal', () => {
  it('renders without crashing', () => {
    const component = renderer.create(<ModalTester />);
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders without footer', () => {
    const component = renderer.create(<ModalTester footer={null} />);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
