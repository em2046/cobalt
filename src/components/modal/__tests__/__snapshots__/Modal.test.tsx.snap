// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders without crashing 1`] = `
<div>
  <div />
  <div>
    <div
      className="cobalt-modal-mask"
    />
    <div
      aria-describedby="dialog2Desc"
      aria-labelledby="dialog2Title"
      className="cobalt-modal-wrap"
      onClick={[Function]}
      role="dialog"
    >
      <div
        className="cobalt-modal"
        onClick={[Function]}
      >
        <div
          className="cobalt-modal-header"
        >
          <div
            className="cobalt-modal-title"
            id="dialog2Title"
          />
          <button
            className="cobalt-modal-close"
            onClick={[Function]}
          >
            <div
              className="cobalt-icon"
            >
              <svg
                fill="currentColor"
                height="1em"
                viewBox="0 0 20 20"
                width="1em"
              >
                close.svg
              </svg>
            </div>
          </button>
        </div>
        <div
          className="cobalt-modal-body"
          id="dialog2Desc"
        >
          Here is content of Modal
        </div>
        <div
          className="cobalt-modal-footer"
        >
          <button
            className="cobalt-button cobalt-button-default"
            onClick={[Function]}
          >
            Cancel
          </button>
          <button
            className="cobalt-button cobalt-button-primary"
            onClick={[Function]}
          >
            OK
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal renders without footer 1`] = `
<div>
  <div />
  <div>
    <div
      className="cobalt-modal-mask"
    />
    <div
      aria-describedby="dialog2Desc"
      aria-labelledby="dialog2Title"
      className="cobalt-modal-wrap"
      onClick={[Function]}
      role="dialog"
    >
      <div
        className="cobalt-modal"
        onClick={[Function]}
      >
        <div
          className="cobalt-modal-header"
        >
          <div
            className="cobalt-modal-title"
            id="dialog2Title"
          />
          <button
            className="cobalt-modal-close"
            onClick={[Function]}
          >
            <div
              className="cobalt-icon"
            >
              <svg
                fill="currentColor"
                height="1em"
                viewBox="0 0 20 20"
                width="1em"
              >
                close.svg
              </svg>
            </div>
          </button>
        </div>
        <div
          className="cobalt-modal-body"
          id="dialog2Desc"
        >
          Here is content of Modal
        </div>
      </div>
    </div>
  </div>
</div>
`;
